Disallow entering discounts higher than price

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14080
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
There's a discrepancy between platforms regarding how we allow discounts to be applied to products, in iOS we allow discounts higher than the product price, while Android does not. 

With this PR we get closer to Android by not allowing these, displaying an error message when a merchant attempts to do so, and disallowing the `Add` button of working until is resolved.

![Simulator Screen Recording - iPhone 15 - 2024-10-01 at 11 22 59](https://github.com/user-attachments/assets/1416be44-68f7-4aca-a65c-adaaf1fb55f6)

There are further UI and logic improvements that can be done in this screen, but I have left them outside this PR to keep a reduced scope (eg: The "remove discount" button appears when there are no discounts saved yet).

Note: There seems to be a lot of unnecessary complexity in these discount-related screens, most likely because we haven't updated the logic from when we stopped handling cases that no longer exist (fees). At some point would be good to separate "discounts" from "legacy fees", and improve the general design of how we apply these.

## Testing
* Go to `Orders` > `+` > `+ Add product` > expand the product and tap on `+ Add discount`
* Observe that if you enter a fixed-type discount higher than the product price:
  * An error message appears
  * The `Add` button is disabled
  * Dismissing the modal does not save changes
 * Repeat the same for a percentage-type discount and observe the same behaviour if the discount is higher than 100%.

| Fixed | Percentage |
|--------|--------|
| ![simulator_screenshot_976DD458-DE2C-4C15-9854-74210830F7AC](https://github.com/user-attachments/assets/634076d9-b793-4574-b981-5122fd8d4051) | ![simulator_screenshot_FB0F5C9E-9987-43EB-A02B-5BCADF5DA1E5](https://github.com/user-attachments/assets/7cbdf80d-23b4-4799-9266-c778d2ffd7d0) | 
---

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Disallow entering discounts higher than price

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
update release notes & internal comment

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/FeeOrDiscountLineDetailsViewModel.swift:
```diff
@@ -273,3 +283,16 @@ extension FeeOrDiscountLineDetailsViewModel {
+     if we attempt to do so, it would happen while the state/view is being modified, leading to undefined behaviour, and/or crashing.
+     In order to avoid this we defer the state update until the current rendering cycle is complete.
+     */
+    func updateDiscountDisallowedState(_ value: Bool) {
+        DispatchQueue.main.async {
+            self.discountValueIsDisallowed = value
+        }
+    }
```
I'm not sure how good practice is this, since we generally send the values directly to the publisher, but I don't see a specific problem with it either. Initially I attempted to use combine to observe changes on the price amount and enable/disable appropriately, but that changes the current implementation enough to open a new can of worms in several places.

joshheald replied:
There's nothing wrong with it, but as mentioned above, you shouldn't need it if you do the validation in response to user input, rather than in response to reading the value.

iamgabrielma committed:
Make var gettable only

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/FeeOrDiscountLineDetailsViewModel.swift:
```diff
@@ -111,6 +111,11 @@ final class FeeOrDiscountLineDetailsViewModel: ObservableObject {
+    /// Captures state when a discount value should be disallowed, for example,
+    /// when the discount entered is higher than the total price of a product.
+    ///
+    @Published private(set) var discountValueIsDisallowed: Bool = false
```
I'd suggest either naming this specific to the reason for why it's disallowed, or making it an enum that explains why with code, rather than a bool.

I think at the moment it's only used for the discount being too high, so let's call it
```suggestion
    @Published private(set) var discountExceedsProductPrice: Bool = false
```

iamgabrielma replied:
Updated! 177128191203b9cbd92abfa18026cb158ab65cb6

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/FeeOrDiscountLineDetailsViewModel.swift:
```diff
@@ -146,6 +151,11 @@ final class FeeOrDiscountLineDetailsViewModel: ObservableObject {
+            updateDiscountDisallowedState(true)
+            return currencyFormatter.formatAmount(priceAfterDiscount) ?? ""
+        }
+        updateDiscountDisallowedState(false)
```
Inside a function named `calculatePriceAfterDiscount`, I'd consider this a side-effect. 

Since `calculatePriceAfterDiscount` is used directly from a view for display, it really ought to be a `var`, ideally a `@Published var`, but at least a computed. I know that's pre-existing, but we should avoid adding a side effect here.

It's potentially confusing that we check whether the input is valid as a response to _displaying_ the discount, rather than as a response to the user entering the discount.

A better trigger might be `FeeOrDiscountLineDetailsViewModel.updateAmount(_:)` and `updatePercentage`, which happens when the `onChangeText` for the input field is called. This would also remove the need for the async block too.

If you move it there, you should probably rename the functions to avoid it being a side effect, e.g. `amountTextChanged` and `percentageTextChanged`, but it's much more of a nitpick at that point.

A complicating factor is that the base price is currently passed in to `FeeOrDiscountLineDetailsViewModel` from the view, as an argument to `calculatePriceAfterDiscount`, which is _weird_ – but it's also passed in when the view model is created, as `baseAmountForPercentage`. That's just the base price, with a re-name, I see no reason that we can't use it for calculating an amount discount as well.

If you do all of that, you can have a much simpler function to check the validity:

```
    func checkDiscountValidity() {
        let priceAfterDiscount = baseAmount - finalAmountDecimal

        discountValueIsDisallowed = priceAfterDiscount < 0
    }
```

Then just call that from your text changed functions.

You're working with some unnecessarily convoluted code here, but I think the PR as it stands makes it more complex, and we don't need to do that.

I've only had a fairly short poke at this, but it appears to work. If I'm missing some detail, apologies and just let me know!

iamgabrielma replied:
Thanks for the guidance! Appreciated 💟 

I had three main issues here that made it trickier-than-expected:
1. The _-unnecessary-_ complexity of it all
2. Being forced to use a function with the price property passed in, directly from the view, for the updates.
3. I was unable to see where `updateAmount` or `updatePercentage` were called. Using both "Show Callers" and the "Call Hierarchy" showed nothing, I see now that this was because these are not directly invoked, but passed through the closure via `.onChange()`, most likely to facilitate decoupling, so we do not need to pass a viewmodel in, but this made it harder to follow its actual usage.

By removing the constraint of (2) as you mentioned, and just using the value we pass in the initializer everything became much simpler, and the tests have become also more understandable now: We pass an initial amount, we update the amount with a discount, and we assert if that's allowed or not.

Main update was done on 177128191203b9cbd92abfa18026cb158ab65cb6 , I've also updated the property name on https://github.com/woocommerce/woocommerce-ios/pull/14083/commits/9dda8ea3a57efd82d8d7eaa92290dc14d72c1947 , and made it a computed var on 6e3c39b99651f63ac385eec09a10227ae9e03e4d

I think there's still some oddities to be addressed in this feature (like how much we need a separation between `baseAmount` and `initialTotal`, or if we still need `lineType` at all) but that's for another day :D

joshheald reviewed (approved):
It tests well... but we could easily make it less complex. See review comments for details, and give me a shout if you want any more detail.

I tested on an iPad running iOS 17.7, no issues related to the different display approach there.

iamgabrielma committed:
Simplify price calculation after discount

By using the existing `baseAmountForPercentage` that we pass to the VM in the init, we can remove the requirement to pass the `price` to the function from the view, which simplifies both the calculation of the discount, as well as validating when a discount shouldn’t be allowed.

iamgabrielma committed:
Update property names. Make compile

iamgabrielma committed:
Update tests

iamgabrielma committed:
Make function into var. Rename

iamgabrielma committed:
remove variable no longer needed

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/FeeOrDiscountLineDetailsViewModel.swift:
```diff
@@ -140,22 +142,21 @@ final class FeeOrDiscountLineDetailsViewModel: ObservableObject {
-        }
-        let discount = NSDecimalNumber(decimal: finalAmountDecimal)
-        let priceAfterDiscount = price.subtracting(discount)
+    var recalculateFormattedPriceAfterDiscount: String {
```
```suggestion
    var formattedPriceAfterDiscount: String {
```

Strange to have a verb in a property name, it makes it seem like a function

iamgabrielma replied:
You're right, updated: https://github.com/woocommerce/woocommerce-ios/pull/14083/commits/795fcff991af23c210d64975dfcfbcb2cd1b04d7

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ProductDiscountView.swift:
```diff
@@ -56,7 +60,7 @@ struct ProductDiscountView: View {
                         Spacer()
                         if let price = viewModel.totalPricePreDiscount {
-                            Text(discountDetailsViewModel.calculatePriceAfterDiscount(price))
+                            Text(discountDetailsViewModel.recalculateFormattedPriceAfterDiscount)
```
```suggestion
                            Text(discountDetailsViewModel.formattedPriceAfterDiscount)
```

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/FeeOrDiscountLineDetailsViewModelTests.swift:
```diff
@@ -411,19 +411,72 @@ final class FeeOrDiscountLineDetailsViewModelTests: XCTestCase {
+        // Simulates the amount the merchant inputs in the discount input field
+        viewModel.updateAmount(discountStringInInputField)
+
+        let discountedPrice = viewModel.recalculateFormattedPriceAfterDiscount
```
```suggestion
        let discountedPrice = viewModel.formattedPriceAfterDiscount
```

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/FeeOrDiscountLineDetailsViewModelTests.swift:
```diff
@@ -411,19 +411,72 @@ final class FeeOrDiscountLineDetailsViewModelTests: XCTestCase {
+        // Simulates the amount the merchant inputs in the discount input field
+        viewModel.updateAmount(discountStringInInputField)
+
+        let discountedPrice = viewModel.recalculateFormattedPriceAfterDiscount
```
```suggestion
        let discountedPrice = viewModel.formattedPriceAfterDiscount
```

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/FeeOrDiscountLineDetailsViewModelTests.swift:
```diff
@@ -411,19 +411,72 @@ final class FeeOrDiscountLineDetailsViewModelTests: XCTestCase {
+        // Simulates the amount the merchant inputs in the discount input field
+        viewModel.updateAmount(discountStringInInputField)
+
+        let discountedPrice = viewModel.recalculateFormattedPriceAfterDiscount
```
```suggestion
        let discountedPrice = viewModel.formattedPriceAfterDiscount
```

joshheald reviewed (approved):
Thanks for the changes, the code looks a lot better. I've suggested a rename in comments.

There is a bug I've found, minor but worth a quick check while you have context – if you enter an invalid discount, switch to the other type ($/%), enter a valid discount there, then switch back to the first type, it will let you add the previously-entered invalid discount.

## Repro
- Create an order and add a product to it
- Expand the product and enter a discount greater than its price
- Change to % discount – observe that the validation message is not cleared, even though the discount shows as $0
- Type any percentage <= 100% – observe that the validation message is cleared
- Change to $ discount – observe that the previous amount is still there
- Tap Add
- You have just discounted by more than the price.

This is happening because we only check the validity when we type something – we could perhaps also do that check when we change the type.

iamgabrielma committed:
rename variable

iamgabrielma committed:
Checks for discount validity on type switch

Discount validity is only checked when the input field changes, but is not checked if a merchant happens to switch between discount types (fixed vs percentual), which can lead to invalid state when a discount is modified in one view, and then they switch type, as discount validity won’t be checked again until the input is updated.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/227">20.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma commented:
> This is happening because we only check the validity when we type something – we could perhaps also do that check when we change the type.

Good catch, you're right on: when we switch between fixed and percentage discounts we recreate a new view instance of `LineDetailView`, This will only check for discount validity once the input field has been modified, so it won't happen immediately after a merchant has switched between discount types, since at that point `onChange(:)` hasn't been called because nothing has been edited.

I've initially tried to fix it in the view model when we perform the discount line type switch, but this ends crashing the app due what seems to be a massive call stack to the computed variable from view recreation. Instead I've just added a call for discount validity check when the view appears, which would happen only once each time we switch between discount types.  Fixed here: https://github.com/woocommerce/woocommerce-ios/pull/14083/commits/b94b56712c5b92f2e775fcf793c516438a24186a

![Simulator Screen Recording - iPhone 15 - 2024-10-03 at 10 27 59](https://github.com/user-attachments/assets/532530c6-4b23-4224-bdfe-640b52aab0d4)

iamgabrielma committed:
Merge branch 'trunk' into task/14080-disallow-negative-product-discount
